### PR TITLE
Revoke token when app is disabled

### PR DESCRIPTION
**Pull request by VivekVinushanth**
### Proposed changes in this pull request
This PR revokes active tokens and authz code, consent when the application got disabled

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAOImpl.java [hunk 1]
```diff
@@ -691,6 +691,66 @@ public void updateAppAndRevokeTokensAndAuthzCodes(String consumerKey, Properties
         }
     }
 
+    public void revokeTokensAndAuthzCodes(String consumerKey, String[] authorizationCodes, String[] accessTokens)
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAOImpl.java:694]: Why don't we introduce two simple methods for the two tasks,
1. revoke tokens
2. revoke authorization codes

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAOImpl.java [hunk 2]
```diff
@@ -691,6 +691,66 @@ public void updateAppAndRevokeTokensAndAuthzCodes(String consumerKey, Properties
...
+            deactivateActiveCodesStatement.setInt(3, appTenantId);
+            deactivateActiveCodesStatement.executeUpdate();
+
+            IdentityDatabaseUtil.commitTransaction(connection);
+
+        } catch (SQLException e) {
+            IdentityDatabaseUtil.rollbackTransaction(connection);
+            throw new IdentityApplicationManagementException("Error while executing the SQL statement.", e);
+        } finally {
+            IdentityDatabaseUtil.closeStatement(revokeActiveTokensStatement);
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAOImpl.java:749]: Let's write the code in the auto closable way.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 3]
```diff
@@ -645,6 +648,73 @@ private void revokeAccessTokensWhenSaaSDisabled(final ServiceProvider servicePro
         }
     }
 
+    /**
+     * Revokes tokens of OAuth applications if application is disabled.
+     *
+     * @param serviceProvider Service Provider
+     * @param tenantDomain    Application tenant domain
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:655]: Complete the java doc comment with the throws part.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 4]
```diff
@@ -645,6 +648,73 @@ private void revokeAccessTokensWhenSaaSDisabled(final ServiceProvider servicePro
...
+     */
+    private void revokeTokensWhenApplicationDisabled(ServiceProvider serviceProvider, String tenantDomain)
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:657]: This method will be called if the application is disabled. So you don't need to denote it from the method name of this private method. Shall we modify the method name just as revokeTokens?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 5]
```diff
@@ -645,6 +648,73 @@ private void revokeAccessTokensWhenSaaSDisabled(final ServiceProvider servicePro
...
+                try {
+                    activeDetailedTokens = OAuthTokenPersistenceFactory
+                            .getInstance().getAccessTokenDAO().getActiveAcessTokenDataByConsumerKey(oauthKey);
+                    String[] accessTokens = new String[activeDetailedTokens.size()];
+                    for (AccessTokenDO detailToken : activeDetailedTokens) {
+                        String token = detailToken.getAccessToken();
+                        accessTokens[countToken] = token;
+                        countToken++;
+                    }
+                    Set<String> authorizationCodes = OAuthTokenPersistenceFactory.getInstance()
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:678]: let's have a separate private method for authorization code revokes

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 6]
```diff
@@ -645,6 +648,73 @@ private void revokeAccessTokensWhenSaaSDisabled(final ServiceProvider servicePro
...
+        }
+    }
+
+    /**
+     * Revokes consent given against OAuth applications when application is disabled.
+     *
+     * @param serviceProvider Service Provider
+     * @param tenantDomain    Application tenant domain
+     */
+    private void revokeConsentWhenApplicationDisabled(final ServiceProvider serviceProvider, final String tenantDomain)
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:700]: Modify the method name by removing the 'WhenApplicationDisabled' part.

**Comment by DMHP** [comment]
@VivekVinushanth Please talk with @Thumimku regarding the token exchange flow. We may need to revoke subject tokens as well.

**Comment by Thumimku** [comment]
How this will handle token exchange for jwt tokens?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java (hunk 5 above)
Comment (VivekVinushanth) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:707]: Resolved with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2484/commits/1f0d6cb338178ec6f1ea45294448c5470696a61f

**Comment by VivekVinushanth** [comment]
> How this will handle token exchange for jwt tokens?

Hope you are referring to the token revocation. I will check upon that.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthUserConsentedScopesDAO.java [hunk 7]
```diff
@@ -107,6 +107,17 @@ void updateExistingConsentForApplication(String userId, String appId, int tenant
     void deleteUserConsentOfApplication(String userId, String appId, int tenantId)
             throws IdentityOAuth2ScopeConsentException;
 
+    /**
+     * Remove user's consent given for an application.
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthUserConsentedScopesDAO.java:111]:
```suggestion
     * Remove users' consent given for applications.
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAO.java [hunk 8]
```diff
@@ -67,6 +67,16 @@ void updateAppAndRevokeTokensAndAuthzCodes(String consumerKey, Properties proper
                                                String[] authorizationCodes, String[] accessTokens)
             throws IdentityOAuth2Exception, IdentityApplicationManagementException;
 
+    default void revokeTokens(String consumerKey, String[] accessTokens)
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAO.java:70]: can we method comment please

Comment (VivekVinushanth) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAO.java:77]: https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2484/commits/4b81c64df3e3c71f8b1715cc02e6ad55e5d8ffe0

**Review by Thumimku — APPROVED** [review]
LGTM,

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAO.java [hunk 9]
```diff
@@ -67,6 +67,30 @@ void updateAppAndRevokeTokensAndAuthzCodes(String consumerKey, Properties proper
                                                String[] authorizationCodes, String[] accessTokens)
             throws IdentityOAuth2Exception, IdentityApplicationManagementException;
 
+    /**
+     * Revoke active access tokens issued against application.
+     *
+     * @param consumerKey    OAuth application consumer key.
+     * @param accessTokens   Active access tokens.
+     * @throws IdentityOAuth2Exception, IdentityApplicationManagementException
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAO.java:75]:
```suggestion
     * @throws IdentityOAuth2Exception IdentityApplicationManagementException
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAO.java [hunk 10]
```diff
@@ -67,6 +67,30 @@ void updateAppAndRevokeTokensAndAuthzCodes(String consumerKey, Properties proper
...
+            throws IdentityOAuth2Exception, IdentityApplicationManagementException {
+
+    }
+
+    /**
+     * Revoke authorize codes issued against application.
+     *
+     * @param consumerKey          OAuth application consumer key.
+     * @param authorizationCodes   Active authorization codes.
+     * @throws IdentityApplicationManagementException
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAO.java:87]:
```suggestion
     * @throws IdentityApplicationManagementException IdentityApplicationManagementException
```

**Review by Thumimku — APPROVED** [review]

**Review by DMHP — APPROVED** [review]

**Pull request merged**